text/0000-arbitrary-self-types-v2.md

```diff
@@ -233,19 +233,27 @@ As noted in the rationale section, the currently nightly implementation implemen
 ## No blanket implementation for `Deref`
 [no-blanket-implementation]: #no-blanket-implementation
 
-The other major approach previously discussed is to have a `Receiver` trait, as proposed in this RFC, but without a blanket implementation for `T: Deref`. Possible advantages:
+The other major approach previously discussed is to have a `Receiver` trait, as proposed in this RFC, but without a blanket implementation for `T: Deref`. Blanket implementations are unusual for core Rust traits, but the authors of this RFC believe it's a powerful tool to reduce user confusion. In particular, a different specification of `Target` for these two traits would sometimes lead the compiler to explore radically different chains of types when determining the candidates for dereferencing and method calls. It's easy to imagine cases where this would be confusing for users. Imagine a type `A` dereferences to `B` yet allows method calls only on `C` (or perhaps more confusingly, allows method calls only on `Pin<&mut B>`).
 
-* It seems more in keeping with Rust norms where types can be in full control of the traits they implement
-* It would allow a type to decide that it wants to be dereferencable, without allowing method calls on the inner type
-* It would allow a type to specify a different `Target` for `Deref` vs `Receiver`.
+We therefore intentionally constrain the use of this feature using a blanket implementation:
```

Comment:
I think it's worth mentioning the instant compatibility with older `Deref` types in third party crates by adding the `Receiver` impl - it is both a good and a bad thing that all types implementing `Deref` will immediately be valid receivers without needing any adjustment to the original crates.

Reply:
I think this is addressed (sort of) in the Method Shadowing section.

Type `B<A>` suddenly becoming a valid receiver for type `A` doesn't actually make any difference to anyone, because there are no behavioral changes unless somebody comes along and types `impl A { fn new_function(self: B<A>) { .. } }`.

Given the need to actually type a new function to use this feature, I don't think that this "instant compatibility" makes a difference in practice?

Reply:
The "instant" compatibility means that the user implementing `A` doesn't need to wait for their upstream `B` to release a new version with `Receiver` implemented. That will speed adoption of `Receiver` a fair bit compared to waiting for ecosystem churn. But probably when considering the final place where Rust will end up, the factor of how long a feature will take to travel through the ecosystem is irrelevant.

Reply:
Ah I see. Thanks.